pkg/serviceaccounts/controllers/pull_secrets.go

```diff
@@ -0,0 +1,320 @@
+
+// PullSecretControllerOptions contains options for the PullSecretController
+type PullSecretControllerOptions struct {
+	// SecretResync is the time.Duration at which to fully re-list secrets.
+	// If zero, re-list will be delayed as long as possible
+	SecretResync time.Duration
+}
+
+// NewPullSecretController returns a new *PullSecretController.
+func NewPullSecretController(cl client.Interface, options PullSecretControllerOptions) *PullSecretController {
+	e := &PullSecretController{
+		client: cl,
+	}
+
+	_, e.secretController = framework.NewIndexerInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return e.client.Secrets(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return e.client.Secrets(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), rv)
+			},
+		},
+		&api.Secret{},
+		options.SecretResync,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    e.secretAdded,
+			DeleteFunc: e.secretDeleted,
+		},
+		cache.Indexers{"namespace": cache.MetaNamespaceIndexFunc},
+	)
+
+	e.secretsSynced = e.secretController.HasSynced
+
+	return e
+}
+
+// The PullSecretController watches for service account tokens to be created or deleted.
+// On create, it creates a corresponding dockercfg secret and attaches it to the service account
+// as both an ImagePullSecret and as a MountableSecret.  On delete, it removes the secret
+// and the references.  It does not watch for changes to the ServiceAccount.  If a user wishes to
+// remove references to the ImagePullSecret, they may freely do so.
+type PullSecretController struct {
+	stopChan chan struct{}
+
+	client client.Interface
+
+	secretController *framework.Controller
+
+	secretsSynced func() bool
+}
+
+// Runs controller loops and returns immediately
+func (e *PullSecretController) Run() {
+	if e.stopChan == nil {
+		e.stopChan = make(chan struct{})
+		go e.secretController.Run(e.stopChan)
+	}
+}
+
+// Stop gracefully shuts down this controller
+func (e *PullSecretController) Stop() {
+	if e.stopChan != nil {
+		close(e.stopChan)
+		e.stopChan = nil
+	}
+}
+
+// secretAdded reacts to a token secret being created by creating a corresponding dockercfg secret if it doesn't already exist.
+func (e *PullSecretController) secretAdded(obj interface{}) {
+	secret := obj.(*api.Secret)
+
+	if _, tokenExists := secret.Data[api.ServiceAccountTokenKey]; secret.Type == api.SecretTypeServiceAccountToken && tokenExists {
+		dockercfgSecret, err := e.findDockercfgSecret(secret)
+		if dockercfgSecret == nil || kapierrors.IsNotFound(err) {
+			// we don't have a dockercfgSecret yet, create one
+			dockercfgSecret, err = e.createDockerPullSecret(secret)
+			if err != nil {
+				glog.Error(err)
+				return
+			}
+		}
+
+		for i := 1; i <= NumServiceAccountUpdateRetries; i++ {
+			if err := e.createDockerPullSecretReference(dockercfgSecret); err != nil {
```

Comment:
If the user removed the service account's references to this dockercfg, this will re-add the references when the controller starts up